polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -1343,6 +1343,130 @@ class RaiseSetNDRange
+
+    // It must be a "kernel" annotation.
+    FailureOr<StringRef> failureOrAnnotationStr = getAnnotation(op);
+    if (failed(failureOrAnnotationStr) || *failureOrAnnotationStr != "kernel")
+      return failure();
+
+    // There should be exactly one store to the marked pointer.
+    Value annotatedPtr = op.getVal();
+    auto [store, lambdaObj] =
+        getUniqueAssignment(annotatedPtr, /*allowMemcpy=*/false);
+    if (!store)
+      return failure();
+
+    auto alloca = dyn_cast_or_null<LLVM::AllocaOp>(lambdaObj.getDefiningOp());
+    if (!alloca || !alloca.getElemType().has_value())
+      return failure();
+
+    auto lambdaClassTy =
+        dyn_cast<LLVM::LLVMStructType>(alloca.getElemType().value());
+    if (!lambdaClassTy)
+      return failure();
+
+    auto captureTypes = lambdaClassTy.getBody();
+
+    // Look for unique assignments that represent the capturing of kernel args.
+    // Assuming the no-op `getelementptr %lambdaObj[0, 0]` was folded, the
+    // first capture is actually a store to the pointer.
+    if (auto [captureOp, capturedVal] = getUniqueAssignment(lambdaObj);
+        captureOp) {
+      capturedVal = tryToBroaden(capturedVal, captureTypes[0]);
+      rewriter.setInsertionPointAfter(captureOp);
+      rewriter.create<sycl::SYCLHostSetCaptured>(captureOp->getLoc(), lambdaObj,
+                                                 rewriter.getI64IntegerAttr(0),
+                                                 capturedVal);
+    }
+
+    // All other captures are unique assignments to GEPs with two constant
+    // indices `[0, <capture #>]` to the lambda object.
+    for (auto *user : lambdaObj.getUsers()) {
+      auto gep = dyn_cast<LLVM::GEPOp>(user);
+      if (!gep || !gep.getElemType().has_value() ||
+          gep.getElemType().value() != lambdaClassTy ||
+          !gep.getDynamicIndices().empty())
+        continue;
+
+      auto indices = gep.getRawConstantIndices();
+      if (indices.size() != 2 || indices[0] != 0)
+        continue;
+      unsigned captureIdx = indices[1];
+
+      auto [captureOp, capturedVal] = getUniqueAssignment(gep);
+      if (!captureOp)
+        continue;
+
+      capturedVal = tryToBroaden(capturedVal, captureTypes[captureIdx]);
+      rewriter.setInsertionPointAfter(captureOp);
+      rewriter.create<sycl::SYCLHostSetCaptured>(
+          captureOp->getLoc(), lambdaObj,
+          rewriter.getI64IntegerAttr(captureIdx), capturedVal);
+    }
+
+    // Finally erase the annotation op.
+    rewriter.eraseOp(op);
+    return success();
+  }
+
+private:
+  /// If there is exactly one store or memcpy to \p ptr, return it and the
+  /// stored value/copied-from pointer; otherwise return `nullptr` and an emtpy
+  /// value.
+  static std::tuple<Operation *, Value>
+  getUniqueAssignment(Value ptr, bool allowMemcpy = true) {
+    Operation *op = nullptr;
+    Value value;
+    for (auto *user : ptr.getUsers())
+      if (auto v = TypeSwitch<Operation *, Value>(user)
```

Comment:
How do we handle the case where `ptr` is casted to another ptr to create an alias, then there are 2 store operations (one to the original ptr and one to the casted ptr) ? 

I think this function would detect that there is a single store (to the original ptr) and report there is a unique assignment to the memory the ptr points to when that isn't the case.

Reply:
true, we should check that there is no address taken possible for all users.

Reply:
Correct, we currently don't handle this situation, in an attempt to keep the raising simple here because it happens on idiomatic, compiler-generated code.

Reply:
True, but right now we're only dealing with kernel lambdas defined in the same C++ statement as the kernel launch itself, so it's unlikely that captured values are re-assigned. The approach in this PR does identify the relevant stores/memcopies, but cannot prove the absence of other writes because there *are* aliasing pointers. For example, capturing of an accessor is detected via an GEP to `%lambda[0, x]` into the lambda object's struct type, but there are also GEPs `%lambda[0, x]` into enclosed accessor struct and again into the enclosed shared_ptr struct. However, we know that they are harmless because they're just part of how the accessor as a whole is copied, if that makes sense.

Reply:
How about we only allow store and gep (and memcpy), and verify that the gep index don't match the input ptr's index? When there exists user of any other kind, then return nullptr.

Reply:
I curated a list of allowed users pf `ptr`, i.e. marker intrinsics, reading from the pointer, and GEPs with constant indices. I also needed to make it possible to allow one specific store of the pointer (because the lambda object is stored into the pointer that is annotated by the frontend). Otherwise this should have reasonably tightened the pattern to our use-case, WDYT?

Reply:
That's good with me.